### PR TITLE
[RJA-680][feature] Environment(instance) lease time

### PR DESCRIPTION
**Pull request by bugzmanov**

### CODE REVIEW

**Comment by bugzmanov** [comment]
https://github.com/griddynamics/OpenGenesis/issues/373

File: backend/src/main/scala/com/griddynamics/genesis/frontend/GenesisRestService.scala [hunk 1]
```diff
@@ -162,6 +175,36 @@ class GenesisRestService(storeService: StoreService,
 
   def stepExists(stepId: Int, envId: Int) = storeService.stepExists(stepId, envId)
 
+  def updateTimeToLive(projectId: Int, envId: Int, timeToLiveMillis: Long): ExtendedResult[Date] = {
+    val env = storeService.findEnv(envId, projectId).getOrElse {
+      return Failure(isNotFound = true, compoundServiceErrors = Seq("Coulnd't find environment %d in project %d".format(envId, projectId)))
```
Comment (ddurnev) [inline backend/src/main/scala/com/griddynamics/genesis/frontend/GenesisRestService.scala]: Couldn't

File: backend/src/main/scala/com/griddynamics/genesis/frontend/GenesisRestService.scala [hunk 2]
```diff
@@ -162,6 +175,36 @@ class GenesisRestService(storeService: StoreService,
...
+    } catch {
+      case e: Exception =>
+        log.error(e, s"Failed to update time to live for environment $envId")
+        Failure(compoundServiceErrors = Seq(s"Failed to update time to live. Cause: ${e.getMessage}"))
+    }
+  }
+
+  def removeTimeToLive(projectId: Int, envId: Int): ExtendedResult[Boolean] = {
+    val env = storeService.findEnv(envId, projectId).getOrElse {
+      return Failure(isNotFound = true, compoundServiceErrors = Seq("Coulnd't find environment %d in project %d".format(envId, projectId)))
```
Comment (ddurnev) [inline backend/src/main/scala/com/griddynamics/genesis/frontend/GenesisRestService.scala]: Could use scala 2.10 placeholders in String here?

File: backend/src/main/scala/com/griddynamics/genesis/scheduler/Executions.scala [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  final def jobKey: JobKey = new JobKey(id, GroupIdentity.forEnv(envId))
+}
+
+sealed trait Execution extends ExecutionId {
+  def toJobDataMap: JobDataMap
+  def jobClass: Class[_ <: Job]
+}
+
+case class WorkflowExecution(workflow: String, envId: Int, projectId: Int) extends Execution {
+  val id = "execution-" + workflow +"-env-" + envId
```
Comment (ddurnev) [inline backend/src/main/scala/com/griddynamics/genesis/scheduler/Executions.scala]: Could use scala 2.10 String interpolation here?

File: backend/src/main/scala/com/griddynamics/genesis/scheduler/jobs/DestructionStatusCheckJob.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import com.griddynamics.genesis.scheduler.{NotificationService, DestructionCheck}
+import com.griddynamics.genesis.configuration.MailServiceContext
+import com.griddynamics.genesis.users.UserService
+
+class DestructionStatusCheckJob(storeService: StoreService,
+                                notificationService: NotificationService) extends Job {
+
+  def execute(context: JobExecutionContext) {
+    val execution = new DestructionCheck(context.getMergedJobDataMap)
+    val env = storeService.findEnv(execution.envId, execution.projectId).get //get
```
Comment (ddurnev) [inline backend/src/main/scala/com/griddynamics/genesis/scheduler/jobs/DestructionStatusCheckJob.scala]: What if get() fails?

File: backend/src/main/scala/com/griddynamics/genesis/scheduler/jobs/WorkflowExecutionJob.scala [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Date
+import java.util.concurrent.TimeUnit
+import com.griddynamics.genesis.api.ExtendedResult
+import com.griddynamics.genesis.scheduler.WorkflowExecution
+
+class WorkflowExecutionJob(broker: RequestBroker, storeService: StoreService) extends Job {
+
+  def execute(context: JobExecutionContext) {
+    val details = new WorkflowExecution(context.getMergedJobDataMap)
+    val env = storeService.findEnv(details.envId, details.projectId).get
```
Comment (ddurnev) [inline backend/src/main/scala/com/griddynamics/genesis/scheduler/jobs/WorkflowExecutionJob.scala]: What if get() fails?

File: frontend/src/main/scala/com/griddynamics/genesis/rest/EnvironmentsController.scala [hunk 6]
```diff
@@ -38,10 +38,18 @@ import com.griddynamics.genesis.api.Failure
...
 import org.springframework.security.access.AccessDeniedException
 import com.griddynamics.genesis.repository.ConfigurationRepository
 import org.apache.commons.lang3.StringEscapeUtils
+import java.util.concurrent.TimeUnit
+import com.griddynamics.genesis.api.ActionTracking
+import com.griddynamics.genesis.api.EnvironmentDetails
+import com.griddynamics.genesis.api.Failure
+import com.griddynamics.genesis.api.WorkflowHistory
+import com.griddynamics.genesis.api.Success
+import com.griddynamics.genesis.api.StepLogEntry
```
Comment (ddurnev) [inline frontend/src/main/scala/com/griddynamics/genesis/rest/EnvironmentsController.scala:51]: import com.griddynamics.genesis.api._ ?

File: frontend/src/main/scala/com/griddynamics/genesis/rest/EnvironmentsController.scala [hunk 7]
```diff
@@ -60,7 +68,7 @@ class EnvironmentsController extends RestApiExceptionsHandler {
   @RequestMapping(value=Array(""), method = Array(RequestMethod.POST))
   @ResponseBody
   def createEnv(@PathVariable("projectId") projectId: Int, request: HttpServletRequest, response : HttpServletResponse): ExtendedResult[Int] = {
-    val paramsMap = extractParamsMap(request)
+    implicit val paramsMap = extractParamsMap(request)
```
Comment (ddurnev) [inline frontend/src/main/scala/com/griddynamics/genesis/rest/EnvironmentsController.scala]: Implicit declared here is used?

File: ui/src/main/resources/genesis/app/templates/createenv/environment_settings.html [hunk 8]
```diff
@@ -23,7 +23,20 @@
...
+            <div style="margin-top: 10px; margin-bottom: 10px;">
+                <label for="time-to-live" class="label" >Automatically destroy</label>
+                <div class="value">
+                   <select id="time-to-live" name="timeToLive">
+                       <option value="">Do not destroy</option>
+                       <option value="60">in 1 hour</option>
+                       <option value="240">in 4 hours</option>
+                       <option value="480">in 8 hours</option>
+                       <option value="1440">in 24 hours</option>
+                       <option value="10080">in 1 weak</option>
```
Comment (ddurnev) [inline ui/src/main/resources/genesis/app/templates/createenv/environment_settings.html]: week

File: ui/src/main/resources/genesis/app/templates/env_details.html [hunk 9]
```diff
@@ -10,6 +10,19 @@
...
+<div id="expand-env-life-time" style="display: none; ">
+    <form>
+        <label for="new-env-ttl" style="display: inline-block; width: 48%">Set instance lifespan: </label>
+        <select id="new-env-ttl" name="new-env-ttl" style="width: 48%">
+            <option value="">unlimited</option>
+            <option value="60">1 hour</option>
+            <option value="240" selected="selected">4 hours</option>
+            <option value="480">8 hours</option>
+            <option value="1440">24 hours</option>
+            <option value="10080">1 weak</option>
```
Comment (ddurnev) [inline ui/src/main/resources/genesis/app/templates/env_details.html]: week

File: backend/src/main/scala/com/griddynamics/genesis/scheduler/jobs/WorkflowExecutionJob.scala (hunk 5 above)
Comment (bugzmanov) [inline backend/src/main/scala/com/griddynamics/genesis/scheduler/jobs/WorkflowExecutionJob.scala]: fixed

File: frontend/src/main/scala/com/griddynamics/genesis/rest/EnvironmentsController.scala (hunk 6 above)
Comment (bugzmanov) [inline frontend/src/main/scala/com/griddynamics/genesis/rest/EnvironmentsController.scala:51]: wont fix

**Pull request merged**